Bindings/Java/Matlab/Hopper_Device/OptimizeHopper.m

```diff
@@ -25,81 +26,184 @@
+xlabel('time (s)')
+ylabel('height')
 
-% Variable bounds
-lb = zeros(numVariables, 1);
-ub = ones(numVariables, 1);
+end
 
-% Initial guess
-x0 = 0.25*ub;
+function f = objective(x, problem)
+% OBJECTIVE
+%   Simulate the hopper with the current controls and report hop height.
 
-% Visualize initial guess
-hopper = buildHopperFromSolution(x0);
-EvaluateHopper(hopper, true, true);
+% See BUILDHOPPERFROMSOLUTION below for details.
+hopper = buildHopperFromSolution(x, problem);
 
-% Optimize!
-opts.TolFun = 0.01;
-[x, f] = fmincon(@objective,x0,[],[],[],[],lb,ub,[],opts);
+% Pass the hopper model to create a forward simulation with the current 
+% guess for the actuator controls. This is the same function called after a
+% hopper model is generated with BuildInteractiveHopperSolution from a 
+% solution created in the InteractiveHopper GUI.
+[peakHeight, ~] = EvaluateHopper(hopper, false, true);
 
-% Evaluate optimized solution
-hopper = buildHopperFromSolution(x);
-EvaluateHopper(hopper, true, true);
+% Negate peak height for minimization.
+f = -peakHeight;
 
 end
 
-function hopper = buildHopperFromSolution(x)
+function hopper = buildHopperFromSolution(x, problem)
 % BUILDHOPPERFROMSOLUTION
-%   Build the hopper with the passive device wrapped and the active device
-%   unwrapped (behind the knee). The optimization variables 'x' determine 
-%   the device properties and the timing of the hop.
-
-import org.opensim.modeling.*;
-
-% These parameters determine the stiffness in the passive device and the
-% maximum force in the active device, both on a 0-100 scale. The range of
-% values for each device correspond to the slider values in the GUI which
-% can be accessed by running InteractiveHopper.m.
-passiveParameter = 100*x(1);
-activeParameter = 100*x(2);
-
-% Construct controls from the last optimization variable, x(3). This
-% variable is the time when the active device is turned off (which 
-% releases the loaded spring) and the vastus muscle is turned to 
-% coordinate a hop.
-% For deviceControl and excitation, the first row contains time nodes
-% and the second row contains control values.
-deviceControl = [0.0 0.01 x(3) x(3)+0.01 5.0;
-                 0.0 1.0  1.0  0.0       0.0];
-excitation =    [0.0      x(3) x(3)+0.01 5.0;
-                 0.0      0.0  1.0       1.0];
+%   Build the hopper for the current iterate. The optimization variables 
+%   'x' determine timing of the hop(s).
+
+% Construct the actuator controls. See CONSTRUCTCONTROLS below for details.
+[deviceControl, muscleExcitation] = constructControls(x, problem);
           
-% Build the hopper. This is the same function called after a solution is
-% is created in the InteractiveHopper GUI. The default muscle is "The
-% Average Joe".
+% Build the hopper with the passive device wrapped and the active device
+% unwrapped (behind the knee). This is the same function called after a
+% solution is created in the InteractiveHopper GUI. Here we use 'The Katie 
```

Comment:
Consider "This is the same function that the The InteractiveHopper GUI calls when you click Simulate." (mainly to avoid saying "solution" for a guess/"iterate").

Reply:
Good idea.